### PR TITLE
HTML parsing, socket for feedback and email notification if wanted

### PR DESCRIPTION
**Pull request by Molyna**
Need to confirm that the parsing everything freely to html is fine (assumption is that the actual ardoq application handles it)

Socket also needs checking with how much data we actually want to transmit and do we want more/less?

### CODE REVIEW

File: resources/form.html [hunk 1]
```diff
@@ -92,7 +92,14 @@
...
           <label>
             <input name="ignorer" type="checkbox"> Ignore validation? (Not recommended!)
-            </label>
+          </label>
+        </div>
+      </div>
+      <div class="form-group">
+        <div class="col-sm-offset-2 col-sm-10 checkbox" id="Notify">
+          <label>
+            <input name="notifier" type="checkbox"> Send you an e-mail upon completion? (Useful for large documents)
```
Comment (expez) [inline resources/form.html]: Tror "Email notification on completion" er bedre copy.

File: src/ardoq/swagger/api.clj [hunk 2]
```diff
@@ -38,22 +40,27 @@
...
-    (swagger/import-swagger client spec url name headers)
+    (do (socket-send "Ignoring validation - Importing Swagger 1" false)
+        (swagger/import-swagger client spec url name headers))
     (let [{:keys [success message]} (validate/validate-swagger "schemav1.json" (generate-string spec))]
       (if success
-        (do (println "Valid Swagger1")
+        (do (socket-send "Valid Swagger - Importing Swagger 1" false)
             (swagger/import-swagger client spec url name headers)
             (println "Done importing swagger doc from " url "."))
         (do (println "Not a valid Swagger file\nError: ")
```
Comment (expez) [inline src/ardoq/swagger/api.clj]: Er det ikke bedre å bruke en logge enn å sprinkle `println` rundt omkring?

File: src/ardoq/swagger/api.clj [hunk 3]
```diff
@@ -68,46 +75,77 @@
...
+  (let [url (str (:url client) "/api/user/notify/email")]
+    (try
+      (->> {:subject "Swagger import failed" :body (str "Your Open API (Swagger) workspace is failed to import.\n" e)} 
+           (generate-string)
+           (assoc (:options client) :body)
+           (http/post url))
+      (catch Exception e
+        (println "Failed to send e-mail")))
+    )
+)
```
Comment (expez) [inline src/ardoq/swagger/api.clj]: litt weird formatting av trailing parens her

File: src/ardoq/swagger/api.clj [hunk 4]
```diff
@@ -68,46 +75,77 @@
...
+           (assoc (:options client) :body)
+           (http/post url))
+      (catch Exception e
+        (println "Failed to send e-mail")))))
+
+(defn notify-failure [session client e]
+  (let [url (str (:url client) "/api/user/notify/email")]
+    (try
+      (->> {:subject "Swagger import failed" :body (str "Your Open API (Swagger) workspace is failed to import.\n" e)} 
+           (generate-string)
```
Comment (expez) [inline src/ardoq/swagger/api.clj]: superfluous "is"

File: src/ardoq/swagger/api.clj [hunk 5]
```diff
@@ -68,46 +75,77 @@
...
+           (try
+             (let [wid (get-spec client url wsname (read-headers headers) swag ignorer)]
+               (socket-close)
+               (when notifier
+                 (notify-success wid org session client))
+               (str (:referer session) "/app/view/workspace/" wid "?org=" org))
+             (catch com.fasterxml.jackson.core.JsonParseException e
+               (notify-failure session client "Failed to parse swagger endpoint")
+               (.printStackTrace e)
+               {:status 406
```
Comment (expez) [inline src/ardoq/swagger/api.clj]: 406 brukes for å indikere feil ved content negotiation.

f.eks hvis du ber om xml i header, server bare har json og ikke tilbyr en default representation.

400 - bad request er bedre her, fordi brukeren ber oss om å jobbe mot et endepunkt vi ikke klarer parse.

File: src/ardoq/swagger/api.clj [hunk 6]
```diff
@@ -68,46 +75,77 @@
       (version2 client spec wsname ignore-validate)
       (version1 client spec url wsname headers ignore-validate))))
 
-(defn swagger-api [{:keys [config]}]
+(defn notify-success [wid org session client]
```
Comment (expez) [inline src/ardoq/swagger/api.clj]: Bedre navn her er `send-{success,failure}-email!` så det kommer frem at vi 1) gjør IO 2) Faktisk sender en mail og ikke f.eks sender melding over ws.

File: src/ardoq/swagger/api.clj [hunk 7]
```diff
@@ -68,46 +75,77 @@
...
+               (notify-failure session client "Failed to parse swagger endpoint")
+               (.printStackTrace e)
+               {:status 406
                 :headers {"Content-Type" "application/json"}
-                :body (json/write-str {:error (str (-> e ex-data :trace-redirects first) " returned 404")})}
+                :body (json/write-str {:error (str "Unable to parse swagger endpoint.")})})
+             (catch IllegalArgumentException e
+               (notify-failure session client "Swagger file does not conform to Swagger specification")
+               (.printStackTrace e)
                {:status 406
```
Comment (expez) [inline src/ardoq/swagger/api.clj]: 400 er også bedre statuskode her

File: src/ardoq/swagger/api.clj [hunk 8]
```diff
@@ -68,46 +75,77 @@
...
-           (catch Exception e
-             (.printStackTrace e)
-             {:status 500
-              :headers {"Content-Type" "application/json"}
-              :body (json/write-str {:error (str "An unexpected error occurred! ")})})))))
+                :body (json/write-str {:error (.getMessage e)})})
+             (catch clojure.lang.ExceptionInfo e
+               (notify-failure session client "An unexpected error occured!")
+               (.printStackTrace e)
+               (if (= 404 (-> e ex-data :status))
```
Comment (expez) [inline src/ardoq/swagger/api.clj]: 404 (not found) brukes bare når klienten ber om en ressurs som serveren ikke finner / kjenner til. 

statuskode 500 (internal server error) brukes for "Noe gikk galt på serveren.  Du får ikke vite hva :p"

Forøvrig finnes det en supernice emacs pakke som heter [know your http well](https://github.com/for-GET/know-your-http-well) som lar deg slå opp hva statuskoder, headers og alt mulig HTTP relatert betyr.  @ebaxt kan sikkert også like den

File: src/ardoq/swagger/swagger.clj [hunk 9]
```diff
@@ -257,7 +261,9 @@
              :operations operations
              :refs refs}]
     (delete-resources client workspace (:apis resource-listing))
+    (socket-send "Removed surplus resources")
```
Comment (expez) [inline src/ardoq/swagger/swagger.clj]: Tror jeg liker "Done pruning resources" og pruning operations under

**Comment by expez** [comment]
Aight dette ser bra ut!

Tror det vil være bedre hvis vi forteller hva vi holder på med istedet for hva vi har gjort.  Så i stedet for i loggen:

```
Created 5 models
Pruned 9 resources
Glorious victory!  Update complete!
```

Så står det 

```
Creating models...
Pruning resources...
Winning! All done!
```

**Comment by expez** [comment]
Funket ikke helt som det skulle.  Mangler secure websocket connection:

image

**Comment by Molyna** [comment]
Updated according to comments. Fixed the minor spelling errors, added a check for secure vs insecure when connecting to the socket and set better http status codes (we now use 400, 404, 422 and 500 for error messages).

The 404 is staying, it is an actual 404. Swagger 1 had a setup where you had the main file, this then pointed to a new document on a new url. If these extra urls throw a 404 we check and then throw our own 404 as a result. 

The 422 is in accordance with what people are doing now for valid but semantically incorrect  content. And in case the swagger file isn't validated I feel that fits with it better than the 400 (it also allows for some different error codes depending on situations). 

I also updated swagger 1 in accordance with how we do swagger 2 in the socket. So the status messages go this way

```
Updating models
Updated 22 models
Updating resources
Updated 30 resources
```

It also tries to report how many it is updating each time, although this isn't possible always without going very spammy (like with swagger 1).

**Comment by Molyna** [comment]
Er denne testet igjen, @expez?

**Comment by expez** [comment]
@Molyna Nei, har ikke gjort det enda.

**Comment by ebaxt** [comment]
@Molyna Still doesn't use wss when called over https it looks like.
screen shot 2016-03-31 at 18 04 25

**Comment by Molyna** [comment]
Latest commit should correctly extract the protocol type and create a socket accordingly.

**Comment by Molyna** [comment]
I'll test this properly myself to make sure we don't get things like yesterday before this pull request goes through.

**Pull request merged**